### DISCUSSION TITLE
Is there any getting started contributing instructions for using Mill?

### DISCUSSION THREAD

**Discussion by lihaoyi** [chipsalliance/chisel#4324] [General]
I tried cloning the repo on my m1 mac, and `./mill __.compile` fails with `jextract` missing. Installing that, it fails with

```
lihaoyi chisel$ PATH="$PATH:/Users/lihaoyi/test/jextract-22/bin" ./mill -i __.compile
[336/679] circtpanamabinding.generatedSources
ERROR: /Users/lihaoyi/Github/chisel/circtpanamabinding/jextract-headers.h:4:10: fatal error: 'mlir-c/BuiltinTypes.h' file not found
1 targets failed
circtpanamabinding.generatedSources os.SubprocessException: Result of jextract…: 4

    os.proc.call(ProcessOps.scala:95)
    millbuild.common$HasJextractGeneratedSources.$anonfun$generatedSources$3(common.sc:155)
```

I haven't managed to figure it out yet, but I thought I'd ask before I try further. Are there any other requirements I need to install to compile stuff? I didn't see anything obvious in the docs, but maybe I missed it

**Comment by sequencer** [comment]
Hello, @lihaoyi, 
There are two ways to get it compiled in mill:
- See GitHub Action CI: https://github.com/chipsalliance/chisel/blob/da986875765f02ed96547952da4b01d1b3f4eab2/.github/workflows/test.yml#L85-L118, you can download `jextract` and `circt`(providing the native binary).
- We use nix to manage the native dependency to make it reproducible: You can use any linux machine which has installed nix to use the flow, see this [link](https://nixos.org/download/). After the installation. `nix develop` can download the required library to provide the header and library which Java can link to.

**Reply by lihaoyi** [reply]
Thanks! As a follow up, is Mac OSX supported? Or do I have to do this all on linux?

**Reply by sequencer** [reply]
> Thanks! As a follow up, is Mac OSX supported? Or do I have to do this all on linux?

I'm not a Darwin user, and I maintain the Nix script in the chisel project. Nix on darwin is a nightmare, so I personally don't have spare time to support it.
But for the native way of downloading circt, setting env, using mill to compile, that should be possible, but I personally doubt the reliability of Panama linking to dylib in Darwin. Feel free to try and issue a bug report:)

**Reply by sequencer** [reply]
BTW, please use JDK 21, since panama api has changed, and chisel only support Java LTS(although Panama api is not stable in this LTS)

**Reply by seldridge** [reply]
> As a follow up, is Mac OSX supported? Or do I have to do this all on linux?

I've built it on macOS w/ MacPorts to install Java 21, setting up JAVA_HOME to point at its install location, and then manually installing `jextract` for Java 21 and putting it's entire contents of `bin/` onto my PATH. (This does include a different `java` binary, but I haven't had issues with this.) After that, everything builds, including the panama pieces.

This is admittedly tedious...

**Comment by jackkoenig** [comment]
Welcome Haoyi! Thank you for your great libraries and tools 🙂 

The Panama aspects of Chisel (which require jextract and other dependencies) are optional[^1] and not required for typical development. I develop Chisel on my M1 Mac with JDK 17 and just don't build those subprojects. I typically just build the `chisel[2.13.14]` project, eg. `./mill chisel[2.13.14].compile`.

[^1]: They are not currently part of the Chisel published artifacts. They will be eventually, but we need to improve the build flow here first (probably by downloading jextract and other needed native dependencies for the developer's platform in Mill).

**Comment by lihaoyi** [comment]
Thanks for all the tips! Will try them out.

Just to give a bit more context here, part of the reason I'm trying to build this codebase is to do a comparative case study of Mill vs SBT, and since Chisel has both build tools that would make the comparison easier (e.g. I won't need to port an existing SBT build). 

If I could ask a favor of you guys, I'd love if you could help write up a short paragraph or two about your experiences with Mill v.s. SBT: why you decided to try Mill, what's good and bad about each tool, etc.. That kind of real-world experience and testimonials would be really valuable

**Reply by sequencer** [reply]
SBT using a opinionated DSL try to implement a dialect that not intuitively understood by most of Scala programmers: the DSL itself is not for the real world build system, but a sugar language that force user to learn as another language. on the other side, mill provides the `Task`-based DSL matching the perfectly what build DAG really is, that's the essentially reason for me to use mill.
